File: test/perl/unit.pl
#!/usr/bin/perl

use lib "$ENV{'MET_TEST_BASE'}/lib";

use strict;
use warnings;

use POSIX;
use VxUtil;
use XML::Parser;
use Data::Dumper;
use Time::HiRes qw( gettimeofday tv_interval );
use Getopt::Long;


sub usage {
  print "usage: $0 [-log {log_file}] [-cmd] [-memchk] [-callchk] [-noexit] {test_xml}

  where:
        log: if present, write output from each test to the specified file
        cmd: if present, print the test commands but do not run them, 
               overrides -log
     memchk: if present, activate valgrind with memcheck
    callchk: if present, activate valgrind with callcheck
     noexit: if present, the unit tester will continue executing subsequent
               tests when a test fails
   test_xml: file containing the unit test(s) to perform

";
}

# parse the input options
my ($mgnc, $mpnc, $file_log, $cmd_only, $noexit, $memchk, $callchk);
GetOptions("log=s" => \$file_log, "cmd" => \$cmd_only, "noexit" => \$noexit,
           "memchk" => \$memchk, "callchk" => \$callchk)
  or die "ERROR: parsing input options";

# parse the input test xml file parameter
my $test_xml = shift or usage() and exit;

# if command  only mode is enabled, disable logging
$cmd_only and $file_log = 0;

# open the log file, if requested
open(my $fh_log, ">", $file_log) or die "ERROR: unable to open log file $file_log\n"
  if $file_log;

# initialize parser object and parse the string
my $parser = XML::Parser->new( Style => "Tree" );
my $tree = eval { $parser->parsefile( $test_xml ); };

# report any error that stopped parsing
if( $@ ){
  $@ =~ s/at \/.*?$//s;
  die "ERROR: parsing error $@\n";
}

# parse the children of the met_test element
$tree->[0] eq "met_test" or die "ERROR: unexpected top-level element " . $tree->[0] . "\n";
my @tests = build_tests( @{ $tree->[1] } );

# determine the max length of the test names
my $name_wid = 12;
$name_wid < length($_->{"name"}) and $name_wid = length($_->{"name"}) for @tests;

# default return value
my $ret_val = 0;
my $VALGRIND_OPT_MEM ="--leak-check=full --show-leak-kinds=all --error-limit=no -v";
my $VALGRIND_OPT_CALL ="--tool=callgrind --dump-instr=yes --simulate-cache=yes --collect-jumps=yes";


# run each test
for my $test (@tests){

  # print the test name
  $cmd_only or printf "TEST: %-*s - ", $name_wid, $test->{"name"};

  # prepare the output space
  my @outputs = ( @{$test->{"out_pnc"}}, @{$test->{"out_gnc"}}, @{$test->{"out_stat"}}, @{$test->{"out_ps"}}, @{$test->{"out_exist"}}, @{$test->{"out_not_exist"}} );
  for my $output ( @outputs ){
    -s $output and qx/rm -rf $output/;
    vx_file_mkdir($output);
  }

  # set the test environment variables
  $ENV{$_} = $test->{"env"}{$_} for keys %{ $test->{"env"} };

  # build the text command
  my ($cmd, $ret_ok, $out_ok) = ($test->{"exec"} . $test->{"param"});
  $cmd =~ s/[ \n\t]+$//m;
  if ($memchk) {
    $cmd = "valgrind ".$VALGRIND_OPT_MEM." ".$cmd;
  }
  elsif ($callchk) {
    $cmd = "valgrind ".$VALGRIND_OPT_CALL." ".$cmd;
  }
  
  # if writing a command file, print the environment and command, then loop
  if( $cmd_only ){
    print "export $_=\'" . $test->{"env"}{$_} . "\'\n" for sort keys %{ $test->{"env"} };
    print "$cmd\n\n";
    next;
  }
  
  # run and time the test command
  my $t_start = [gettimeofday];
  my @cmd_outs = qx{$cmd 2>&1};
  my $t_elaps = tv_interval( $t_start );
  unshift @cmd_outs, "$cmd\n";

  # check the return status and output files
  $ret_ok = (! $?);
  if( $ret_ok ){
    $out_ok = 1;
    do{ push @cmd_outs, qx/$mpnc -v $_/; $? and $out_ok = 0 } for ( @{ $test->{"out_pnc"} } );
    do{ push @cmd_outs, qx/$mgnc -v $_/; $? and $out_ok = 0 } for ( @{ $test->{"out_gnc"} } );
    for my $output ( @{ $test->{"out_stat"} } ){
      -s $output or 
        push @cmd_outs, "ERROR: stat file missing \"$output\"\n" and $out_ok = 0 and next;
      int(qx{cat $output 2>/dev/null | egrep -v '^VERSION' | wc -l}) or
        push @cmd_outs, "ERROR: stat data missing from file \"$output\"\n" and $out_ok = 0;
    }
    for my $output ( @{ $test->{"out_ps"} } ){
      -s $output or 
        push @cmd_outs, "ERROR: postscript file missing \"$output\"\n" and $out_ok = 0 and next;
      !(qx{gs -sDEVICE=nullpage -dQUIET -dNOPAUSE -dBATCH $output 2>/dev/null}) or
        push @cmd_outs, "ERROR: ghostscript error for postscript file \"$output\"\n" and $out_ok = 0;
    }
    for my $output ( @{ $test->{"out_exist"} } ){
      -s $output or 
        push @cmd_outs, "ERROR: file missing when it should exist \"$output\"\n" and $out_ok = 0 and next;
    }
    for my $output ( @{ $test->{"out_not_exist"} } ){
      !(-s $output) or 
        push @cmd_outs, "ERROR: file exists when it should be missing \"$output\"\n" and $out_ok = 0 and next;
    }
    
  }

  # print the test result
  printf "%s - %7.3f sec\n", $ret_ok && $out_ok ? "pass" : "FAIL", sprintf("%7.3f", $t_elaps);

  # build a list of environment variable exports for reporting
  my @envs;
  push @envs, "export $_=\'" . $test->{"env"}{$_} . "\'\n" for sort keys %{ $test->{"env"} };

  # if the log file is activated, print the test information
  if( $file_log ){
    print $fh_log "\n\n";
    print $fh_log "$_" for (@envs, @cmd_outs);
    print $fh_log "\n\n";
  }

  # on failure, print the problematic test and exit, if requested
  if( !($ret_ok && $out_ok) ){
    print "$_" for (@envs, @cmd_outs);
    $noexit or exit 1;
    print "\n\n";
  }

}



#######################################################################
# build_tests()
#
#   This function assumes that the inputs are the body elements of
#   a parsed XML file using XML::Parser.  The components of each test
#   element are parsed and the test elements are returned as an array
#   of hashes.
#
#   Arguments:
#      xml_test = array of body elements containing test structures
#
#######################################################################

sub build_tests {

  # ignore the attributes of the top-level element
  shift;

  # parse each test
  my ($test_idx, @tests) = (1);
  while( @_ ){

    # examine the next element
    my $elm_test = shift;
    $elm_test eq 'test_dir'
      and $mgnc = ${$_[0]}[2] . "/bin/mgnc.sh"
      and $mpnc = ${$_[0]}[2] . "/bin/mpnc.sh"; 
    $elm_test ne 'test' and shift and next;

    # parse the test child components
    my @childs = @{ shift @_ };

    # extract the test name from the attributes
    my %test;
    $test{"name"} = $childs[0]->{'name'}
      or die "ERROR: name attribute not found for test ", $test_idx++ . "\n";
    shift @childs;

    # extract the test details from the test child elements
    while( @childs ){

      my $test_child = shift @childs;
      $test_child or shift @childs and next;

      # set the exec or parm string
      $test_child eq "exec" || $test_child eq "param"
        and $test{$test_child} = repl_env($childs[0][2]) 
        and shift @childs and next;

      # build the output file arrays
      if( $test_child eq "output" ){
        local @_ = @{ $childs[0] };
        shift and shift @childs;

        my (@out_pncs, @out_gncs, @out_stats, @out_pss, @out_exists, @out_not_exists);
        while( @_ ){
          my $out_child = shift;
          $out_child or shift and next;
          "point_nc"  eq $out_child and push @out_pncs,       repl_env($_[0][2]);
          "grid_nc"   eq $out_child and push @out_gncs,       repl_env($_[0][2]);
          "stat"      eq $out_child and push @out_stats,      repl_env($_[0][2]);   
          "ps"        eq $out_child and push @out_pss,        repl_env($_[0][2]);
          "exist"     eq $out_child and push @out_exists,     repl_env($_[0][2]);
          "not_exist" eq $out_child and push @out_not_exists, repl_env($_[0][2]);          
          shift;
        }

        @test{ ("out_pnc",  "out_gnc",   "out_stat",  "out_ps",  "out_exist",  "out_not_exist") } = 
               (\@out_pncs, \@out_gncs,  \@out_stats, \@out_pss, \@out_exists, \@out_not_exists);
      }

      # build the environment map
      elsif( $test_child eq "env" ){
        local @_ = @{ $childs[0] };
        shift and shift @childs;

        my %env;
        while( @_ ){
          my $env_child = shift;
          $env_child or shift and next;

          # parse the pair name and value
          if( "pair" eq $env_child ){
            my @pair_childs = @{ $_[0] };
            shift @pair_childs;
            my ($pair_name, $pair_value);
            while( @pair_childs ){
              my $pair_child = shift @pair_childs;
              $pair_child or shift @pair_childs and next;
              "name"  eq $pair_child and $pair_name  = $pair_childs[0][2];
              "value" eq $pair_child and $pair_value = $pair_childs[0][2];
            }
            $pair_name or die "ERROR: env pair in test \"" . $test{"name"} . "\" " .
                              "missing name or value\n";            
            $env{$pair_name} = defined($pair_value) ? $pair_value : "";
          }
          $test{"env"} = \%env;
        }
      }

    }

    # verify the structure of the test element
    $test{"exec"}   or die "ERROR: test " . $test{"name"} . " missing exec element\n";
    $test{"param"}  or die "ERROR: test " . $test{"name"} . " missing param element\n";
    ( $test{"out_pnc"} && $test{"out_gnc"} && $test{"out_stat"} && $test{"out_ps"} && $test{"out_exist"} && $test{"out_not_exist"} ) 
                    or die "ERROR: test " . $test{"name"} . " missing output element\n";

    # add the test to the list of parsed tests
    push @tests, \%test;

  }  # END: while( @_ )

  return @tests;
}

#######################################################################
# repl_env()
#
#   This function assumes that the inputs are strings containing one or
#   more placeholders for environment variable values with syntax
#   ${ENV_NAME}.  The entire placeholder is replaced with the value of
#   the specified environment variable.  If no named variable exists,
#   repl_env() dies with error message. 
#
#   Arguments:
#     repl_list = array of strings containing environment value place-
#                 holders to be replaced
#
#######################################################################

sub repl_env {
  @_ or return 0;
  my @ret;
  while( @_ ){
    local $_ = shift;
    while( /\$\{(\w+)\}/ ){
      my $val = $ENV{$1} or die "ERROR: environment variable $1 not found\n";
      s/\$\{$1\}/$val/g;
    }
    push @ret, $_;
  }
  return wantarray ? @ret : $ret[0];
}

File: internal/test_unit/python/unit.py
#! /usr/bin/env python3

from datetime import datetime as dt
import logging
import os
from pathlib import Path
import re
import subprocess
import sys
import xml.etree.ElementTree as ET

def unit(test_xml, file_log=None, cmd_only=False, noexit=False, memchk=False, callchk=False):
    """
    Parse a unit test xml file, run the associated tests, and display test results.

    Parameters
    -----------
    test_xml : pathlike
        path to file containing the unit test(s) to perform
    file_log : pathlike, default None
        if present, write output from each test to the specified file
    cmd_only : bool, default False
        if true, print the test commands but do not run them (overrides file_log)
    noexit : bool, default False
        if true, the unit tester will continue executing subsequent
                tests when a test fails
    memchk : bool, default False
        if true, activate valgrind with memcheck
    callchk : bool, default False
        if true, activate valgrind with callcheck
    """
  
    # initialize logger
    logger = logging.getLogger(__name__)
    logger.setLevel(logging.DEBUG)

    # create/add console handler
    ch = logging.StreamHandler()
    ch.setLevel(logging.INFO)
    logger.addHandler(ch)

    # create/add file handler
    if file_log and not cmd_only:
        fh = logging.FileHandler(file_log)
        fh.setLevel(logging.DEBUG)
        logger.addHandler(fh)
   
    # parse xml file
    try:
        test_root = ET.parse(test_xml)
    except Exception as e:
        logger.exception(f"ERROR: Unable to parse xml from {test_xml}")
        raise

    # parse the children of the met_test element
    if test_root.getroot().tag != 'met_test':
        logger.error(f"ERROR: unexpected top-level element. Expected 'met_test', got '{test_root.tag}'")
        sys.exit(1)
    else:
        # read test_dir
        try:
            test_dir = test_root.find('test_dir').text
            mgnc = repl_env(test_dir + '/bin/mgnc.sh')
            mpnc = repl_env(test_dir + '/bin/mpnc.sh')
        except Exception as e:
            logger.warning(f"WARNING: unable to read test_dir from {test_xml}")
            pass

        tests = build_tests(test_root)

    # determine the max length of the test names
    #   not used, unless format of test result display is changed
    name_wid = max([len(test['name']) for test in tests])
    
    VALGRIND_OPT_MEM ="--leak-check=full --show-leak-kinds=all --error-limit=no -v"
    VALGRIND_OPT_CALL ="--tool=callgrind --dump-instr=yes --simulate-cache=yes --collect-jumps=yes"

    # run each test
    for test in tests:
    #   # print the test name ... may want to change this to only if cmd_only=False
        logger.debug("\n")
        logger.info(f"TEST: {test['name']}")

    #   # prepare the output space
        output_keys = [key for key in test.keys() if key.startswith('out_')]
        outputs = [output for key in output_keys for output in test[key]]
        for output in outputs:
            try:
                Path(output).unlink()
            except FileNotFoundError:
                pass
            except Exception as e:
                logger.exception()
                raise
            output_dir = Path(output).parent
            output_dir.mkdir(parents=True, exist_ok=True)     #should error/warning be raised if dir already exists? 

    #   # set the test environment variables
        set_envs = []
        if 'env' in test.keys():
            for key, val in sorted(test['env'].items()):
                os.environ[key] = val
                set_cmd = f"export {key}={val}"
                logger.debug(set_cmd)
                set_envs.append(set_cmd)

    #   # build the text command
        cmd = test['exec'] + test['param']
        cmd = re.sub('[ \n\t]+$', '', cmd)  # not sure this is doing what it should;
                                            #  may need to remove the +$ from the regex?
        if memchk:
            cmd = f"valgrind {VALGRIND_OPT_MEM} {cmd}"
        elif callchk:
            cmd = f"valgrind {VALGRIND_OPT_CALL} {cmd}"

        
    #   # if writing a command file, print the environment and command, then loop
        #   consider tying this into logging...
        if cmd_only:
            if 'env' in test.keys():
                for key, val in sorted(test['env'].items()):
                    print(f"export {key}={val}")
            print(f"{cmd}")
            if 'env' in test.keys():
                for key, val in sorted(test['env'].items()):
                    print(f"unset {key}")
            print("\n")
    
    #   # run and time the test command
        else:
            logger.debug(f"{cmd}")
            t_start = dt.now()
            cmd_return = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True, shell=True)
            t_elaps = dt.now() - t_start

            cmd_outs = cmd_return.stdout
            logger.debug(f"{cmd_outs}")
            logger.debug(f"Return code: {cmd_return.returncode}")

        #   # check the return status and output files
            ret_ok = not cmd_return.returncode
            if ret_ok:
                out_ok = True

                for filepath in test['out_pnc']:
                    result = subprocess.run([mpnc, '-v', filepath], 
                                            stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
                    cmd_outs += ("\n"+result.stdout)
                    logger.debug(result.stdout)
                    if result.returncode:
                        out_ok = False
                
                for filepath in test['out_gnc']:
                    result = subprocess.run([mgnc, '-v', filepath], 
                                            stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True)
                    cmd_outs += ("\n"+result.stdout)
                    logger.debug(result.stdout)
                    if result.returncode:
                        out_ok = False
                
                for filepath in test['out_stat']:
                    # check stat file exists and is nonzero size
                    try:
                        filesize = os.stat(filepath).st_size
                        if filesize==0:
                            cmd_outs += (f"\nERROR: stat file empty {filepath}\n")
                            out_ok = False
                            break
                    except FileNotFoundError:
                        cmd_outs += (f"\nERROR: stat file missing {filepath}\n")
                        logger.debug(result.stdout)
                        out_ok = False
                        break
                    # check stat file has non-header lines
                    with open(filepath) as f:
                        numlines = len([l for l in f.readlines() if not l.startswith('VERSION')])
                    if numlines==0:
                        cmd_outs += (f"\nERROR: stat data missing from file {filepath}\n")
                        out_ok = False

                for filepath in test['out_ps']:
                    # check postscript file exists and is nonzero size
                    try:
                        filesize = os.stat(filepath).st_size
                        if filesize==0:
                            cmd_outs += (f"\nERROR: postscript file empty {filepath}\n")
                            out_ok = False
                            break
                    except FileNotFoundError:
                        cmd_outs += (f"\nERROR: postscript file missing {filepath}\n")
                        out_ok = False
                        break
                    # check for ghostscript errors
                    result = subprocess.run(['gs', '-sDEVICE=nullpage', '-dQUIET', '-dNOPAUSE', '-dBATCH', filepath])
                    if result.returncode:
                        cmd_outs += (f"\nERROR: ghostscript error for postscript file {filepath}")
                        out_ok = False
                
                for filepath in test['out_exist']:
                    # check output file exists and is nonzero size
                    try:
                        filesize = os.stat(filepath).st_size
                        if filesize==0:
                            cmd_outs += (f"\nERROR: file empty {filepath}\n")
                            out_ok = False
                            break
                    except FileNotFoundError:
                        cmd_outs += (f"\nERROR: file missing when it should exist {filepath}\n")
                        out_ok = False

                for filepath in test['out_not_exist']:
                    # check output file doesn't exist
                    if os.path.isfile(filepath):
                        cmd_outs += (f"\nERROR: file exists when it should be missing {filepath}\n")
                        out_ok = False

            #   # unset the test environment variables
            unset_envs = []
            if 'env' in test.keys():
                for key, val in sorted(test['env'].items()):
                    del os.environ[key]
                    unset_cmd = f"unset {key}"
                    logger.debug(unset_cmd)
                    unset_envs.append(unset_cmd)
            
            #   # print the test result
            test_result = "pass" if (ret_ok and out_ok) else "FAIL"
            logger.info(f"\t- {test_result} - \t{round(t_elaps.total_seconds(),3)} sec")

            #   # on failure, print the problematic test and exit, if requested
            if not (ret_ok and out_ok):
                logger.info("\n".join(set_envs) + cmd + cmd_outs + "\n".join(unset_envs) + "\n")
                if not noexit:
                    sys.exit(1)


def build_tests(test_root):
    """
    Parse the test components.

    Take an ElementTree element extracted from a unit test xml file.
    Return a list of all tests, where each test is represented as a dictionary,
    with its keys representing each test component.

    Parameters
    ----------
    test_root : ElementTree element
        parsed from XML file containing the unit test(s) to perform

    Returns
    -------
    test_list: 
        list of test dicts, containing test attributes parsed from xml object

    """

    # define logger
    logger = logging.getLogger(__name__)

    # find all tests in test_xml, and create a dictionary of attributes for each test
    test_list = []
    for test_el in test_root.iter('test'):
        test = {}
        try:
            test['name'] = test_el.attrib['name']
        except KeyError:
            logger.error("ERROR: name attribute not found for test")
            raise
            # should just fail this one test and not fully exit?

        for el in test_el:
            if (el.tag=='exec' or el.tag=='param'):
                test[el.tag] = repl_env(el.text)
            elif el.tag=='output':
                test['out_pnc'] = []
                test['out_gnc'] = []
                test['out_stat'] = []
                test['out_ps'] = []
                test['out_exist'] = []
                test['out_not_exist'] = []
                output_names = {
                    'point_nc' : 'out_pnc',
                    'grid_nc' : 'out_gnc',
                    'stat' : 'out_stat',
                    'ps' : 'out_ps',
                    'exist' : 'out_exist',
                    'not_exist' : 'out_not_exist',
                }
                for output_el in el:
                    test[output_names[output_el.tag]].append(repl_env(output_el.text))

            elif el.tag=='env':
                env_dict = {}
                for env_el in el:
                    try:
                        env_name = env_el.find('name').text
                        env_dict[env_name] = env_el.find('value').text
                        if not env_dict[env_name]:
                            env_dict[env_name] = ''
                    except AttributeError:
                        logger.error(f"ERROR: env pair in test \\{test['name']}\\ missing name or value")
                        raise
                        # should just fail this one test and not fully exit?                      

                test['env'] = env_dict

        #   validate test format/details
        expected_keys = ['exec', 'param', 'out_pnc', 'out_gnc', 'out_stat', 'out_ps', 
                         'out_exist', 'out_not_exist']
        for key in expected_keys:
            if key not in test.keys():
                logger.error(f"ERROR: test {test['name']} missing {key} element")
                sys.exit(1)
        
        test_list.append(test)

    return test_list


def repl_env(string_with_ref):
    """
    Take a string with a placeholder for environment variable with syntax
    ${ENV_NAME} and replace placeholder with corresponding value of environment
    variable.

    Parameters
    ----------
    string_with_ref : str
        A string, generally path-like, that includes substring ${ENV_NAME}

    Returns
    -------
    string_with_ref : str
        The provided string with ${ENV_NAME} replaced by corresponding environment variable
    """
    # define logger
    logger = logging.getLogger(__name__)

    envar_ref_list = re.findall('\$\{\w+}', string_with_ref)
    envar_ref_unique = [
        envar_ref_list[i] for i in list(range(len(envar_ref_list))) if (
            envar_ref_list[i] not in envar_ref_list[:i])]

    if len(envar_ref_unique)>0:
        for envar_ref in envar_ref_unique:
            envar_name = envar_ref[2:-1]
            envar = os.getenv(envar_name)
            if not envar:
                logger.error(f"ERROR: environment variable {envar_name} not found")
            string_with_ref = string_with_ref.replace(envar_ref, envar)

    return string_with_ref

if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Run a unit test.")
    parser.add_argument('test_xml', )
    parser.add_argument('-log', metavar='log_file',
                        help='if present, write output from each test to log_file')
    parser.add_argument('-cmd', action='store_true',
                        help='if present, print the test commands but do not run them, overrides -log')
    parser.add_argument('-memchk', action='store_true',
                        help='if present, activate valgrind with memcheck')
    parser.add_argument('-callchk', action='store_true',
                        help='if present, activate valgrind with callcheck')
    parser.add_argument('-noexit', action='store_true',
                        help='if present, the unit tester will continue executing subsequent tests when a test fails')
    args = parser.parse_args()

    unit(test_xml=args.test_xml, file_log=args.log, cmd_only=args.cmd, noexit=args.noexit, memchk=args.memchk, callchk=args.callchk)


